numpy/_core/include/numpy/ndarraytypes.h

```diff
@@ -1796,6 +1796,144 @@ typedef struct {
                            */
 } PyArrayInterface;
 
+/****************************************
+ * NpyString
+ *
+ * Types and macros used by the NpyString API.
+ ****************************************/
+
+/*
+ * A "packed" encoded string. The string data must be accessed by first unpacking the string.
+ */
+typedef struct npy_packed_static_string npy_packed_static_string;
+
+/*
+ * An unpacked read-only view onto the data in a packed string
+ */
+typedef struct npy_unpacked_static_string {
+    size_t size;
+    const char *buf;
+} npy_static_string;
+
+/*
+ * Handles heap allocations for static strings.
+ */
+typedef struct npy_string_allocator npy_string_allocator;
+
+typedef struct {
+    PyArray_Descr base;
+    // The object representing a null value
+    PyObject *na_object;
+    // Flag indicating whether or not to coerce arbitrary objects to strings
+    int coerce;
+    // Flag indicating the na object is NaN-like
+    int has_nan_na;
+    // Flag indicating the na object is a string
+    int has_string_na;
+    // If nonzero, indicates that this instance is owned by an array already
+    int array_owned;
+    // The string data to use when a default string is needed
+    npy_static_string default_string;
+    // The name of the missing data object, if any
+    npy_static_string na_name;
+    // The mutex locking the allocator
+    PyThread_type_lock *allocator_lock;
+    // the allocator should only be directly accessed after
+    // acquiring the allocator_lock and the lock should
+    // be released immediately after the allocator is
+    // no longer needed
+    npy_string_allocator *allocator;
+} StringDTypeObject;
+
+/*
+ * Acquire the mutex locking the allocator attached to *descr*.
+ */
+static inline npy_string_allocator *
+NpyString_acquire_allocator(StringDTypeObject *descr)
+{
+    if (!PyThread_acquire_lock(descr->allocator_lock, NOWAIT_LOCK)) {
+        PyThread_acquire_lock(descr->allocator_lock, WAIT_LOCK);
+    }
+    return descr->allocator;
+}
+
+/*
+ * Simultaneously acquire the mutexes locking the allocators attached to
+ * *descr1* and *descr2*.
+ */
+static inline void
+NpyString_acquire_allocator2(StringDTypeObject *descr1,
+                             StringDTypeObject *descr2,
+                             npy_string_allocator **allocator1,
+                             npy_string_allocator **allocator2)
+{
+    *allocator1 = NpyString_acquire_allocator(descr1);
+    if (descr1 != descr2) {
+        *allocator2 = NpyString_acquire_allocator(descr2);
+    }
+    else {
+        *allocator2 = *allocator1;
+    }
+}
+
+/*
+ * Simultaneously acquire the mutexes locking the allocators attached to
+ * *descr1*, *descr2*, and *descr3*.
+ */
+static inline void
+NpyString_acquire_allocator3(StringDTypeObject *descr1,
```

Comment:
Almost makes me wonder if we should have `NpyString_acquire_allocators(int n, StringDTypeObject *descrs[], npy_string_allocator *out_allocators[])` for anything beyond 2.
OTOH, I guess checking identity on all of them may end up tedious, so probably not really tha tuseful.

Reply:
Much more important point: Making these macros hardcodes the `allocator` and `allocator_lock` slots on the dtype.  That part I suspect makes no problem.

However, you mentioned potentially changing how the lock is implemented, and this hard-codes `PyThread_acquire_lock(descr->allocator_lock, NOWAIT_LOCK)` as *public* API.
Are you confident that `PyThread_aqcuire_lock` is reasonable here?  Because unless you are, `NpyString_acquire_allocator` needs to be a function for the public API.

Reply:
Ah, that's a good point! Thanks so much for being careful about the C API - this is the first time I've tried to write a C API for wide consumption like this and very much appreciate your experience.

I initially thought about putting the lock in the allocator struct. Since the allocator struct is opaque in the public API, I don't need to worry about people poking at the internals before the allocator is acquired, which is the main reason I didn't have it that way in the prototype. I think I'll just move the lock into the allocator struct and make the acquire/release macros take the allocator.

I could also make the `allocator_lock` member a `void *` pointer, but it's simpler if users only need to worry about an `allocator` pointer so let's do that first.

Reply:
I ended up doing this. One wrinkle is that since the allocator struct is opaque, I had to make these acquire/release functions API functions instead of `static inline` header-only functions. I initially was hesitant to do that for performance reasons but on second thought that doesn't matter because we're already paying the indirection penalty for making the allocator struct opaque every time we call a function that takes an allocator as an argument so doing it two more times outside of the loop kernel won't matter much.